File: core/optimizer/semantic_optimizer_engine.py
from __future__ import annotations

from typing import Any, Dict, List


MAX_OPTIMIZATION_PASSES = 10


def optimize_semantic_ir(
    ir: Dict[str, Any],
) -> Dict[str, Any]:

    optimizations: List[Dict[str, Any]] = []

    if "execution_paths" in ir:

        paths = ir["execution_paths"].get("paths", [])

        deduped = []

        seen = set()

        for p in paths:

            key = tuple(p)

            if key not in seen:
                deduped.append(p)
                seen.add(key)

        optimizations.append({
            "type": "deduplicate_execution_paths",
            "before": len(paths),
            "after": len(deduped),
        })

        ir["execution_paths"]["paths"] = deduped

    return {
        "optimized_ir": ir,
        "optimizations": optimizations[
            :MAX_OPTIMIZATION_PASSES
        ],
        "deterministic": True,
    }

File: src/optimizer/semanticOptimizerEngine.ts
/**
 * Converted from Python: core/optimizer/semantic_optimizer_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_OPTIMIZATION_PASSES: any = 10;
export function optimizeSemanticIr(ir: any): any {
  var optimizations: any[] = [];
  if (py.contains(ir, "execution_paths")) {
    var paths: any = py.get(py.at(ir, "execution_paths"), "paths", []);
    var deduped: any[] = [];
    var seen: Set<any> = new Set();
    var p: any;
    for (p of py.iter(paths)) {
      var key: any = [...py.iter(p)];
      if (!py.contains(seen, key)) {
        py.listAppend(deduped, p);
        py.setAdd(seen, key);
      }
    }
    py.listAppend(optimizations, {"type": "deduplicate_execution_paths", "before": py.len(paths), "after": py.len(deduped)});
    py.setItem(py.at(ir, "execution_paths"), "paths", deduped);
  }
  return {"optimized_ir": ir, "optimizations": py.slice(optimizations, null, MAX_OPTIMIZATION_PASSES), "deterministic": true};
}
